subsys/bluetooth/audio/has.c

```diff
@@ -601,77 +845,189 @@ static int bt_has_cp_generic_update(struct has_client *client, const struct has_
 	}
 
-	if (read_presets_req_pending_cp(client)) {
-		const struct has_preset *preset = NULL;
-		bool is_last = true;
+	if (client->read_presets_req.num_presets > 1) {
+		const struct has_preset *next = NULL;
 
-		preset_foreach(client->read_presets_req.start_index, BT_HAS_PRESET_INDEX_LAST,
-			       preset_found, &preset);
+		preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST, preset_found, &next);
 
-		if (unlikely(preset == NULL)) {
-			(void)bt_has_cp_read_preset_rsp(client, NULL, 0x01);
+		is_last = next == NULL;
+	}
 
-			return;
-		}
+	err = bt_has_cp_read_preset_rsp(client, preset, is_last);
+	if (err != 0) {
+		return err;
+	}
 
-		if (client->read_presets_req.num_presets > 1) {
-			const struct has_preset *next = NULL;
+	if (preset->index > client->context->last_preset_index_known) {
+		update_last_preset_index_known(client, preset->index);
+	}
 
-			preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
-				       preset_found, &next);
+	if (!is_last) {
+		client->read_presets_req.start_index = preset->index + 1;
+		client->read_presets_req.num_presets--;
 
-			is_last = next == NULL;
+	atomic_set_bit(client->context->flags, FLAG_PENDING_READ_PRESET_RESPONSE);
+		notify_work_reschedule(client, K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+	}
 
-		}
+	return 0;
+}
 
-		err = bt_has_cp_read_preset_rsp(client, preset, is_last);
-		if (err) {
-			LOG_ERR("bt_has_cp_read_preset_rsp failed (err %d)", err);
-		}
+static int bt_has_cp_preset_record_deleted(struct has_client *client, uint8_t index)
+{
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(struct bt_has_cp_hdr) +
+			      sizeof(struct bt_has_cp_preset_changed) + sizeof(uint8_t));
 
-		if (err || is_last) {
-			read_presets_req_free(client);
-		} else {
-			client->read_presets_req.start_index = preset->index + 1;
-			client->read_presets_req.num_presets--;
-		}
-	} else if (atomic_test_and_clear_bit(client->flags, FLAG_CONTROL_POINT_NOTIFY)) {
-		const struct has_preset *preset = NULL;
-		const struct has_preset *next = NULL;
-		bool is_last = true;
+	preset_changed_prepare(&buf, BT_HAS_CHANGE_ID_PRESET_DELETED, BT_HAS_IS_LAST);
+	net_buf_simple_add_u8(&buf, index);
 
-		preset_foreach(client->preset_changed_index_next,
-			       BT_HAS_PRESET_INDEX_LAST, preset_found, &preset);
+	if (client != NULL) {
+		return control_point_send(client, &buf);
+	} else {
+		return control_point_send_all(&buf);
+	}
+}
 
-		if (preset == NULL) {
-			return;
-		}
+static int preset_list_changed_generic_update_tail(struct has_client *client)
+{
+	const struct has_preset *prev;
+	struct has_preset last = {
+		.index = client->context->last_preset_index_known,
+		.properties = 0x00,
+		.name = " ",
+	};
+	int err;
 
-		preset_foreach(preset->index + 1, BT_HAS_PRESET_INDEX_LAST,
-			       preset_found, &next);
+	prev = preset_get_tail();
 
-		is_last = next == NULL;
+	err = bt_has_cp_generic_update(client, prev ? prev->index : 0x00, last.index,
```

Comment:
`properties = 0x00` seems to have a special meaning. Do we have a define for the `0x00` property? Else suggest to add that.